internal/provider/function_direxists.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/hashicorp/terraform-plugin-framework/function"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+var _ function.Function = &DirectoryExistsFunction{}
+
+type DirectoryExistsFunction struct{}
+
+func NewDirectoryExistsFunction() function.Function {
+	return &DirectoryExistsFunction{}
+}
+
+func (f *DirectoryExistsFunction) Metadata(ctx context.Context, req function.MetadataRequest, resp *function.MetadataResponse) {
+	resp.Name = "direxists"
+}
+
+func (f *DirectoryExistsFunction) Definition(ctx context.Context, req function.DefinitionRequest, resp *function.DefinitionResponse) {
+	resp.Definition = function.Definition{
+		Summary: "Determines whether a directory exists at a given path.",
+		Description: "Given a path string, will return true if the directory exists. " +
+			"This function works only with directories. If used with a regular file, FIFO, or other " +
+			"special mode, it will return an error.",
+
+		Parameters: []function.Parameter{
+			function.StringParameter{
+				Name:        "path",
+				Description: "Relative or absolute path to check for the existence of a directory",
+			},
+		},
+		Return: function.BoolReturn{},
+	}
+}
+
+func (f *DirectoryExistsFunction) Run(ctx context.Context, req function.RunRequest, resp *function.RunResponse) {
+	var directoryPath string
+
+	resp.Diagnostics.Append(req.Arguments.Get(ctx, &directoryPath)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	fi, err := os.Stat(directoryPath)
```

Comment:
One thing that `fileexists()` does that is not done here is automatically expand `~` into the calling user's home directory. I'm not sure we necessarily need to do that in this function, but if that is something we want for consistency, then it is probably best to do that upfront to not introduce a potentially breaking change later.

Reply:
Does d993daa match what you were suggesting? I saw that `fileexists` uses a set base directory and an external `go-homedir` library to expand that path. I believe this commit should behave in the same way

Reply:
Hmm. I don't think stdlib does anything with regards to `~` expansion unless that's changed in newer versions of Go? In general though, the absolute path translation and cleaning is practitioner a benefit here as it should help with troubleshooting whatever relative/funky paths given actually mean to the provider logic if the file system calls generate an error.

Reply:
Option 3 seems reasonable to me in this situation. I might encourage some upfront form of that `pathexpand()` mentioning documentation around this since it will differ from the native `fileexists()` behavior that practitioners might already be using/familiar. 👍

Reply:
Cool, I made some explicit comment and added an example in 782da26543ee4b0c3dc89c66e563437037b40f28 ⚡